stdlib/public/Concurrency/TaskSleep.swift

```diff
@@ -0,0 +1,144 @@
+  /// the sleep completed.
+  private static func onSleepCancel(
+      _ wordPtr: UnsafeMutablePointer<Builtin.Word>,
+      _ continuation: UnsafeContinuation<(), Error>
+  ) {
+    // Indicate that we've finished by putting a "2" into the flag word.
+    let (_, won) = Builtin.cmpxchg_seqcst_seqcst_Word(
+        wordPtr._rawValue,
+        UInt(0)._builtinWordValue,
+        UInt(2)._builtinWordValue)
+
+    if Bool(_builtinBooleanLiteral: won) {
+      // We recorded the task cancellation before the sleep finished, so
+      // invoke the continuation with a the cancellation error.
+      continuation.resume(throwing: _Concurrency.CancellationError())
+    }
+  }
+
+  /// Suspends the current task for _at least_ the given duration
+  /// in nanoseconds, unless the task is cancelled. If the task is cancelled,
+  /// throws \c CancellationError.
+  ///
+  /// This function does _not_ block the underlying thread.
+  public static func sleep(nanoseconds duration: UInt64) async throws {
+    // If the task was already cancelled, go ahead and throw now.
+    try checkCancellation()
+
+    // Allocate storage for the flag word and continuation.
+    let wordPtr = UnsafeMutablePointer<Builtin.Word>.allocate(capacity: 2)
+
+    // Initialize the flag word to 0, which means the continuation has not
+    // executed.
+    Builtin.atomicstore_seqcst_Word(
+        wordPtr._rawValue, UInt(0)._builtinWordValue)
+
+    // A pointer to the storage continuation. Also initialize it to zero, to
+    // indicate that there is no continuation.
+    let continuationPtr = wordPtr + 1
+    Builtin.atomicstore_seqcst_Word(
+      continuationPtr._rawValue, UInt(0)._builtinWordValue)
+
+    do {
+      // Install a cancellation handler to resume the continuation by
+      // throwing CancellationError.
+      try await withTaskCancellationHandler {
+        let _: () = try await withUnsafeThrowingContinuation { continuation in
+          // Stash the continuation so the cancellation handler can see it.
+          Builtin.atomicstore_seqcst_Word(
+            continuationPtr._rawValue,
+            unsafeBitCast(continuation, to: Builtin.Word.self))
+
+          // Create a task that resumes the continuation normally if it
+          // finishes first. Enqueue it directly with the delay, so it fires
+          // when we're done sleeping.
+          let sleepTaskFlags = taskCreateFlags(
+            priority: nil, isChildTask: false, copyTaskLocals: false,
+            inheritContext: false, enqueueJob: false,
+            addPendingGroupTaskUnconditionally: false)
+          let (sleepTask, _) = Builtin.createAsyncTask(sleepTaskFlags) {
+            onSleepWake(wordPtr, continuation)
+          }
+          _enqueueJobGlobalWithDelay(
+              duration, Builtin.convertTaskToJob(sleepTask))
+        }
+      } onCancel: {
+        let continuationWord = continuationPtr.pointee
+        if UInt(continuationWord) != 0 {
```

Comment:
is this safe from here on out if another thread cancels just after the `checkCancellation` at line 84?

Reply:
So, there's an existing race in `withTaskCancellationHandler` that also gets replicated here. When we address that (@ktoso was looking into it), this will also be fixed.

Reply:
I'll get to it this week. I think we'll end up having to "check again" after having stored a cancellation record basically.